pallets/simple-staking/src/lib.rs

```diff
@@ -0,0 +1,182 @@
+		/// Event documentation should end with an array that provides descriptive names for event
+		/// parameters. [something, who]
+		NewInvulnerables(Vec<T::AccountId>),
+		NewMaxAuthorCount(u32),
+		NewAuthorBond(BalanceOf<T>),
+		AuthorAdded(T::AccountId, BalanceOf<T>),
+		AuthorRemoved(T::AccountId)
+	}
+
+	// Errors inform users that something went wrong.
+	#[pallet::error]
+	pub enum Error<T> {
+		MaxAuthors,
+		Unknown,
+		Permission,
+		AlreadyAuthor,
+		NotAuthor
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(_: T::BlockNumber) -> Weight {
+			let author = <authorship::Module<T>>::author();
+			let treasury = T::TreasuryAddress::get();
+			let reward = T::Currency::free_balance(&treasury).div(2u32.into());
+			T::Currency::transfer(&treasury, &author, reward, KeepAlive);
+			// TODO double check: weight 3 reads (author, treasury, balance) 1 writes (transfer)
+			T::DbWeight::get().reads_writes(3, 1)
+		}
+		//TODO on init (or finalize) add to aura set at next era
+	}
+
+	// Dispatchable functions allows users to interact with the pallet and invoke state changes.
+	// These functions materialize as "extrinsics", which are often compared to transactions.
+	// Dispatchable functions must be annotated with a weight and must return a DispatchResult.
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+
+
+		#[pallet::weight(10_000)]
+		pub fn set_invulnerables(origin: OriginFor<T>, new: Vec<T::AccountId>) -> DispatchResultWithPostInfo {
+			T::UpdateOrigin::ensure_origin(origin)?;
+			<Invulnerables<T>>::put(&new);
+			Self::deposit_event(Event::NewInvulnerables(new));
+			Ok(().into())
+		}
+
+		#[pallet::weight(10_000)]
+		pub fn set_max_author_count(origin: OriginFor<T>, max_authors: u32) -> DispatchResultWithPostInfo {
+			T::UpdateOrigin::ensure_origin(origin)?;
+			<MaxAuthors<T>>::put(&max_authors);
+			Self::deposit_event(Event::NewMaxAuthorCount(max_authors));
+			Ok(().into())
+
+		}
+
+
+		#[pallet::weight(10_000)]
+		pub fn set_author_bond(origin: OriginFor<T>, author_bond: BalanceOf<T>) -> DispatchResultWithPostInfo {
+			T::UpdateOrigin::ensure_origin(origin)?;
+			<AuthorBond<T>>::put(&author_bond);
+			Self::deposit_event(Event::NewAuthorBond(author_bond));
+			Ok(().into())
+
+		}
+
+		#[pallet::weight(10_000)]
+		pub fn register_as_author(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+			// lock deposit to start or require min?
+			let who = ensure_signed(origin)?;
+			let deposit = <AuthorBond<T>>::get();
+			let length = <Authors<T>>::decode_len().unwrap_or_default();
+			ensure!((length as u32) < MaxAuthors::<T>::get(), Error::<T>::MaxAuthors);
+			let new_author = AuthorInfo {
+				who: who.clone(),
+				deposit,
+				last_block: None
+			};
+			<Authors<T>>::try_mutate(|authors| -> DispatchResult {
+				if authors.into_iter().any(|author| author.who == who) {
+					Err(Error::<T>::AlreadyAuthor)?
+				} else {
+					T::Currency::reserve(&who, deposit)?;
```

Comment:
Without ensuring that the free balance is `>= deposit`, this is vulnerable to an edge case where the free balance is zero, because [this returns Ok](https://github.com/paritytech/substrate/blob/13bea77c6910f76ef7fa59fec347874c99b2170c/frame/balances/src/lib.rs#L1553).

It's actually quite feasible to call with a zero-balance account, e.g. an anon proxy.